ratatool-diffy/src/main/scala/com/spotify/ratatool/diffy/AvroDiffy.scala

```diff
@@ -20,67 +20,79 @@ package com.spotify.ratatool.diffy
-      x.get(f)
+  // scalastyle:off cyclomatic.complexity method.length
+  private def diff(x: Option[GenericRecord], y: Option[GenericRecord], root: String)
+  : Seq[Delta] = {
+    val schemaFields = (x, y) match {
+      case (Some(xVal), None) => xVal.getSchema.getFields.asScala.toList
+      case (_, Some(yVal)) => yVal.getSchema.getFields.asScala.toList
+      case _ => List()
     }
 
-    y.getSchema.getFields.asScala.flatMap { f =>
+    schemaFields.flatMap { f =>
       val name = f.name()
       val fullName = if (root.isEmpty) name else root + "." + name
       getRawType(f.schema()).getType match {
         case Schema.Type.RECORD =>
-          val a = x.get(name).asInstanceOf[GenericRecord]
-          val b = y.get(name).asInstanceOf[GenericRecord]
-          if (a == null && b == null) {
-            Nil
-          } else if (a == null || b == null) {
-            Seq(Delta(fullName, Option(a), Option(b), UnknownDelta))
-          } else {
-            diff(a, b, fullName)
+          val a = x.flatMap(r => Option(r.get(name).asInstanceOf[GenericRecord]))
+          val b = y.flatMap(r => Option(r.get(name).asInstanceOf[GenericRecord]))
+          (a, b) match {
+            case (None, None) => Nil
+            case (Some(_), None) => Seq(Delta(fullName, a, None, UnknownDelta))
+            case (None, Some(_)) => Seq(Delta(fullName, None, b, UnknownDelta))
+            case (Some(_), Some(_)) => diff(a, b, fullName)
           }
         case Schema.Type.ARRAY if unordered.contains(fullName) =>
           if (f.schema().getElementType.getType == Schema.Type.RECORD
-              && unordered.exists(_.startsWith(s"$fullName."))
-              && unorderedFieldKeys.contains(fullName)) {
-            val a = sortList(x.get(name).asInstanceOf[java.util.List[GenericRecord]],
-              unorderedFieldKeys.get(fullName).map(getField))
-            val b = sortList(y.get(name).asInstanceOf[java.util.List[GenericRecord]],
-              unorderedFieldKeys.get(fullName).map(getField))
-            a.asScala.zip(b.asScala).flatMap{case (l, r) =>
-              diff(l.asInstanceOf[GenericRecord], r.asInstanceOf[GenericRecord], fullName)
-            }.toList
+            && unordered.contains(fullName)
+            && unorderedFieldKeys.contains(fullName)) {
+            val l = x.flatMap(r =>
+              Option(r.get(name).asInstanceOf[java.util.List[GenericRecord]].asScala.toList))
+              .getOrElse(List())
+              .flatMap(r => Try(r.get(unorderedFieldKeys(fullName))).toOption.map(k => (k, r)))
+              .toMap
+            val r = y.flatMap(r =>
+              Option(r.get(name).asInstanceOf[java.util.List[GenericRecord]].asScala.toList))
+              .getOrElse(List())
+              .flatMap(r => Try(r.get(unorderedFieldKeys(fullName))).toOption.map(k => (k, r)))
+              .toMap
+            (l.keySet ++ r.keySet).flatMap(k => diff(l.get(k), r.get(k), fullName)).toList
```

Comment:
Seems like we're saying the diff of two records `x` and `y` that are like `Array[NestedRecord]` is computing by taking any exact matches of some subfield `x.a` and `y.a` of `NestedRecord` and comparing?  

I worry this could not be what some users would expect - e.g. if the arrays are equal size and all subfields are numeric but all slightly different, there would be no way to compute the diff.

Reply:
Synced offline.  

We agreed this does change the behavior since we no longer sort on `unorderedFieldKeys` in a way that'll mark all records as "unkonwn delta" if we want to compare on numeric fields that could be different.  But all use cases we know of are for strings from what I was told.  

Docs should be updated though, esp to clarify that `unorderedFieldKeys` is no longer a sorting key as the docs c currently say.

Reply:
Might also be good to write a test w/ this new behavior.  

The test would also aid as documentation for expected behavior.